src/PowerShell/PowerShellManagerPool.cs

```diff
@@ -1,116 +1,118 @@
-            string invocationId = request.InvocationRequest?.InvocationId;
-
-            // If the pool has an idle one, just use it.
-            if (!_pool.TryTake(out psManager))
-            {
-                // The pool doesn't have an idle one.
-                if (_poolSize < _upperBound &&
-                    Interlocked.Increment(ref _poolSize) <= _upperBound)
-                {
-                    // If the pool hasn't reached its bounded capacity yet, then
-                    // we create a new item and return it.
-                    var logger = new RpcLogger(_msgStream);
-                    logger.SetContext(requestId, invocationId);
-                    psManager = new PowerShellManager(logger);
-
-                    RpcLogger.WriteSystemLog(string.Format(PowerShellWorkerStrings.LogNewPowerShellManagerCreated, _poolSize.ToString()));
-                }
-                else
-                {
-                    // If the pool has reached its bounded capacity, then the thread
-                    // should be blocked until an idle one becomes available.
-                    psManager = _pool.Take();
-                }
-            }
-
-            // Register the function with the Runspace before returning the idle PowerShellManager.
-            FunctionMetadata.RegisterFunctionMetadata(psManager.InstanceId, functionInfo);
-            psManager.Logger.SetContext(requestId, invocationId);
-            return psManager;
-        }
-
-        /// <summary>
-        /// Return a used PowerShellManager instance to the pool.
-        /// </summary>
-        internal void ReclaimUsedWorker(PowerShellManager psManager)
-        {
-            if (psManager != null)
-            {
-                // Unregister the Runspace before reclaiming the used PowerShellManager.
-                FunctionMetadata.UnregisterFunctionMetadata(psManager.InstanceId);
-                psManager.Logger.ResetContext();
-                _pool.Add(psManager);
-            }
-        }
-    }
-}
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Functions.PowerShellWorker.Messaging;
+using Microsoft.Azure.Functions.PowerShellWorker.Utility;
+using Microsoft.Azure.WebJobs.Script.Grpc.Messages;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
+{
+    using System.Management.Automation;
+
+    /// <summary>
+    /// The PowerShellManager pool for the in-proc concurrency support.
+    /// </summary>
+    internal class PowerShellManagerPool
+    {
+        private readonly int _upperBound;
+        private readonly MessagingStream _msgStream;
+        private readonly BlockingCollection<PowerShellManager> _pool;
+        private int _poolSize;
+
+        internal int UpperBound => _upperBound;
```

Comment:
maybe a small comment saying "this is the concurrency upper bound" so it's clear